source/extensions/filters/common/auth/attribute_context.cc

```diff
@@ -0,0 +1,91 @@
+#include "extensions/filters/common/auth/attribute_context.h"
+
+#include "common/http/utility.h"
+#include "common/network/utility.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Filters {
+namespace Common {
+namespace Auth {
+
+void AttributeContextUtils::setSourcePeer(envoy::service::auth::v2alpha::AttributeContext& context,
+                                          const Network::Connection& connection,
+                                          const std::string& service) {
+  setPeer(*context.mutable_source(), connection, service, false);
+}
+
+void AttributeContextUtils::setDestinationPeer(
+    envoy::service::auth::v2alpha::AttributeContext& context,
+    const Network::Connection& connection) {
+  setPeer(*context.mutable_destination(), connection, "", true);
+}
+
+void AttributeContextUtils::setPeer(envoy::service::auth::v2alpha::AttributeContext_Peer& peer,
+                                    const Network::Connection& connection,
+                                    const std::string& service, const bool local) {
+
+  auto addr = local ? connection.localAddress() : connection.remoteAddress();
+  Envoy::Network::Utility::addressToProtobufAddress(*addr, *peer.mutable_address());
+
+  if (!service.empty()) {
+    peer.set_service(service);
+  }
+
+  auto ssl = const_cast<Ssl::Connection*>(connection.ssl());
+  if (ssl) {
+    std::string principal = local ? ssl->uriSanLocalCertificate() : ssl->uriSanPeerCertificate();
+    if (principal.empty()) {
+      principal = local ? ssl->subjectLocalCertificate() : ssl->subjectPeerCertificate();
+    }
+    peer.set_principal(principal);
+  }
+}
+
+void AttributeContextUtils::setHttpRequest(
+    envoy::service::auth::v2alpha::AttributeContext& context,
+    const Envoy::Http::StreamDecoderFilterCallbacks* callbacks,
+    const Envoy::Http::HeaderMap& headers) {
+
+  auto req = context.mutable_request()->mutable_http();
+  auto sdfc = const_cast<Envoy::Http::StreamDecoderFilterCallbacks*>(callbacks);
+
+  auto start = ProtobufUtil::TimeUtil::MicrosecondsToTimestamp(
+      sdfc->requestInfo().startTime().time_since_epoch().count());
+  context.mutable_request()->mutable_time()->MergeFrom(start);
+
+  auto proto = sdfc->requestInfo().protocol();
+  if (proto) {
+    req->set_protocol(Envoy::Http::Utility::getProtocolString(proto.value()));
+  }
+
+  req->set_id(std::to_string(sdfc->streamId()));
+  req->set_size(sdfc->requestInfo().bytesReceived());
+
+  req->set_method(getHeaderStr(headers.Method()));
+  req->set_path(getHeaderStr(headers.Path()));
+  req->set_host(getHeaderStr(headers.Host()));
+  req->set_scheme(getHeaderStr(headers.Scheme()));
+  // TODO(rodaine): add query & fragment fields
+
+  auto mutable_headers = req->mutable_headers();
+  headers.iterate(
```

Comment:
I understand why this is being done for the external auth code, but I'm just throwing it out there now that this type of copy will not be acceptable for the inline auth code. There is no reason to copy all the headers just to look through them again. So I'm just throwing this out there now in case that changes how you want to do things.

(In general we won't do a layer of indirection that is not needed for the inline code). Given that, does this shared code add value? Just trying to understand the future intent here.

Reply:
Hm. yeah that makes sense. Early discussions about RBAC about consuming a `AttributeContext` directly, which would require this conversion. If we want to avoid that, we'd need to create a similar `AuthContext` that just holds references to the connection, request info, and headers. This `AuthContext` could be shared between RBAC/ext_authz, and expose a method to populate the proto message for ext_authz's use. Thoughts?

Reply:
TBH I would abandon this change and just start writing the filter. I don't see the value in copying to an intermediary when the check logic is trivial and can be done inline in a nice way. If you find there is coding sharing possible when you go do that, that's the time to do it IMO.

Reply:
I understand the point to avoid copying data to an intermediary. This totally makes sense. But the benefit here is that the code is more modularized, and potentially more reusable. 

The goal of using AttributeContext is to standardize the interface to any authorization module. If RBAC is implemented as a standalone library, it can potentially be reused in other environment, like shared between different filters, or used in a pure gRPC environment where there is no filter concept. 

Also, the AttributeContext could be filled directly from connection/request info (like being done in this PR), or some fields could be filled by an authentication filter in the future. IMO, it just makes the interface cleaner.

Reply:
@liminw let's cross this bridge when we come to it. This feature is so simple IMO that I really think we are greatly overcomplicating it. I would like to get the filter written since we need it at Lyft and then we can always come back and refactor/share code as needed if it makes sense (given that it's an internal implementation detail and I think we did a great job of getting the proto right).

Reply:
@mattklein123 Sounds good!